### DISCUSSION TITLE
Query regarding user-defined submodel classes

### DISCUSSION THREAD

**Discussion by RhtMht** [pybamm-team/PyBaMM#3787] [Q&A] [answered]
Hi PyBaMM community,

I have experience of few years in the field of cell modeling but am new to python and pybamm. I wish to modify some constant parameters in the existing models (thermal and sei) to function parameters. For example, in the code attached below, I am trying to make the convective heat transfer coefficient a function of cell and ambient temperatures (basically, Grashof and Prandtl numbers). For achieving this, I updated a copy of the lumped class in thermal model and used it instead of the inbuilt thermal models. However, I am getting errors stating " 'ParameterValues' object has no attribute ..." when building the model.  Can you explain what is the issue and how to resolve this?

[code.txt](https://github.com/pybamm-team/PyBaMM/files/14103184/code.txt)

Thank you

**Comment by TomTranter** [comment, marked as answer]
The problem is you are passing paramater_values dict not the lithium_ion.parameters definitions which are symbolic.
```
model_thermal={"thermal": Lumped(model_TDFN.param, model_TDFN.options)}
```
I ran your code and changed the above line then ran into some more errors related to the equations you have defined

**Reply by RhtMht** [reply]
Hi Tom,

Thanks for pointing out the issue. I have implemented this correction in my code. However, I am now facing a keyerror in my code. 

I am trying to change the parameter "Total heat transfer coefficient [W.m-2.K-1]" (which is presently a constant in PyBaMM) to a function of another parameter function "Ambient temperature [K]" and the cell's volume averaged temperature which is the temperature obtained by solving the lumped thermal model. 

To achieve this, I made following changes-
1. In my main code, I have defined a class for lumped thermal model (by modifying the inbuilt lumped thermal model class). In this, I changed the self.param.h_total to self.param.h_total(T_vol_av,T_amb) in line 84.
2. In the inbuilt lithium_ion_parameters file, I removed the line self.h_total = self.therm.h_total (in line 64 of inbuilt file) and added a method h_total to make the convective heat transfer coefficient a function of ambient and cell temperatures (lines 143-145 of modified file).

However, when I try to run my "main_code" with modified "lithium_ion_parameters" file, I get the keyerror issue.

[lithium_ion_parameters.txt](https://github.com/pybamm-team/PyBaMM/files/14122749/lithium_ion_parameters.txt)
[main_code.txt](https://github.com/pybamm-team/PyBaMM/files/14122750/main_code.txt)

Thank you

**Reply by RhtMht** [reply]
Hi Tom, I recognized my mistake. I was using "^" to define exponents instead of the pow function. After correction, the error I am getting is "The residual routine or the linear setup or solve routine had a recoverable error, but IDACalcIC was unable to recover." I have checked that for a temperature difference up to 100 degrees between cell and ambient the convective heat transfer function returns a value between 0-10. The model can run if I return any such values directly and even with more simple functions, but is not running with the function I need to use. Attached is the corrected main file.

[main_code_v02.txt](https://github.com/pybamm-team/PyBaMM/files/14124661/main_code_v02.txt)

Thank you
